Explore/docker builds (#7)
DeiC-HPC/BitBulldozersLab: docker_builds/README.md

> # Docker build pipeline for 4 different container approaches on LUMI

Reviewer: The top level folder name (currently docker_builds) should match the title of the BitBulldozer

- **Keywords:** MPI, MPICH, libfabric, libcxi, container, open source, RCCL, docker, apptainer, OSU
- **Date:** 2025-03-07

# Description

In this section of the repository we explore the different options for building a docker image without proprietary libraries. 
This is based on the LUMI docker build files found here:

> https://github.com/sfantao/lumi-containers/tree/lumi-sep2024

Reviewer: This link 404s

Author: ugh..... the branch is gone.

The main aim is to be able to build docker base images without needing a system similar to LUMI.  
Currently there are open source images that can be used but the interconnects don't work properly. 

# Summary
This repository compares four different container approaches to a native approach for the OSU benchmark suit. 
1) a container only approach.
2) a container approach with full bind mounts from LUMI.

> 3) a container approach where we replace only libfabric + lower level libraries (libcxi)

Reviewer: What does "full" refer to? is it both libfabric, lower level libraries and ROCm?
What does `replace` refer to? should it be bind mount?

Author: Clarified in next commit.

4) a fully open source container

For approaches 2) to 4) we achieve similar performance to a native approach. 

# Issues
- no AWS-OFI-RCCL2 for Libfabric 2.0 containers
- Small message sizes in the fully open source container perform worse than native. 

# Instructions to reproduce:

> ## Requirements:

Reviewer: It would be nice to have at least the original version numbers of the packages used. Also add singularity / apptainer versions.

Author: done in new commit

- Docker 
- Python packages:
  - docker for python (pip install docker)
  - rich (pip install rich)

## Step-by-Step:

> 

Reviewer: ```suggestion
0) These builds were originally done on Raxos / Primebox.
```

1) navigate to docker_builds folder
2) `sudo python3 build_docker.py` builds **3** different docker containers (**WARNING: Takes up to 5 hours on my laptop**) named:
   - `lumi_images:base_image_libfabric1152_mpich314`
   - `lumi_images:base_image_libfabric1152_mpich423`
   - `lumi_images:base_image_libcxi_libfabric1220_mpich423`
3) Each container has to be converted to an apptainer container via `sudo apptainer build $TARGET $SOURCE`
   - where $TARGET and $SOURCE have to be:
     - `base_image_mpich314_libfabric1152.sif` & `docker-daemon:lumi_images:base_image_libfabric1152_mpich314`
     - `base_image_mpich314_libfabric1152.sif` & `docker-daemon:lumi_images:base_image_libfabric1152_mpich423`

>      - `base_image_mpich314_libfabric1152.sif` & `docker-daemon:lumi_images:base_image_libcxi_libfabric1220_mpich423` 

Reviewer: ```suggestion
     - `base_image_mpich423_libfabric1152.sif` & `docker-daemon:lumi_images:base_image_libfabric1152_mpich423`
     - `base_image_mpich423_libfabric1220.sif` & `docker-daemon:lumi_images:base_image_libcxi_libfabric1220_mpich423` 
```

Author: resolved in new commit

> 4) Each of the sif files has to be copied over to lumi via your preferred method e.g. `scp base_image_mpich314_libfabric1152.sif /project/project_XXXXXX/`

Reviewer: ```suggestion
4) Each of the sif files has to be copied over to lumi via your preferred method e.g. `scp base_image_mpich314_libfabric1152.sif /project/project_XXXXXXXXX/LUMI-$USER/`
```

5) Copy all the .sh scripts in the 'tests' folder. (Optional: native folder)
6) Optional: Build native osu benchmark suit. See instructions [here](#native)
7) `sbatch run_XXXXXX.sh TargetImage` (i.e. `sbatch run_base_bandwidth_and_latency_tests.sh base_image_mpich314_libfabric1152.sif`)
8) Look at the resulting txt files (i.e. `base_bandwidth_host_host.txt`). 

# Note:
- https://github.com/apptainer/apptainer/issues/282 
  - There is an issue on Apptainer that would auto find and mount the host MPI libraries into a container. However, no work has been done on this since 2022 and the issue is still open.
- Finding the correct MPI libraries to bindmount into a container. Checkout https://github.com/E4S-Project/e4s-cl
  - The e4s-cl seems like an interesting tool. It can be installed via Spack. 
  - Its essentially a launcher for MPI workloads with containers. It can probably be used with Apptainer and it seems to be helpful for getting all required MPI libraries? 
  - https://e4s-project.github.io/e4s-cl.html 
  - Demo looks super easy

> # TODO List && Research Questions

Reviewer: ```suggestion
# Research Questions
```

Author: done

Questions we aim to answer with this BitBulldozer

- [X] Basic container:
  - [X] bandwidth
  - [X] latency
  - [X] RCCL
- [X] Native:
  - [X] bandwidth
  - [X] latency
  - [X] RCCL
- [X] Lumi bind mount container:
  - [X] bandwidth
  - [X] latency
  - [X] GTL linking:
    - [X] is there a latency difference?
    - [X] Is there a bandwidth difference?
    - [X] Does GTL linking impact NCCL?
- [X] Libfarbic hybrid:
  - [X] bandwidth? 
  - [X] latency?
  - [X] Linking GTL with libfabric Hybrid setup?
    - [X] is there a latency difference?
    - [X] Is there a bandwidth difference?
- [X] Full Open Source
  - [X] bandwidth? 
  - [X] latency?
  - [X] Linking GTL with libfabric Hybrid setup?
    - [X] is there a latency difference?
    - [X] Is there a bandwidth difference?

***
***

> # Compatability:

Reviewer: ```suggestion
# Results Matrix
```

> |                                      | pure container | Full bind mount | libfabric-hybrid | Opensource | Built by default |

Reviewer: What does "Built by default" refer to? Is it just the three containers built in the step-by-step in the README? If that is the case I think the **Y** is confusing and mixed together with Y from the legend. If I have understood the "Built by default" correctly, I think it is sufficient to put a star in before the name such as " * libfabric 1152 & mpich 314" and then in the legend write that * refers to containers build during this README.

| libfabric 1211 & mpich 343           |       (Y)      | (Y)             | N                | ??         | N                |
| libfabric 1220 & mpich 343           |       (Y)      | (Y)             | N                | N          | N                |
| libfabric 1152 & mpich 423           |       (Y)      | Y               | Y                | N          | **Y**            |
| libfabric 1211 & mpich 423           |       (Y)      | (Y)             | N                | (Y)        | N                |
| libfabric 1220 & mpich 423           |       (Y)      | (Y)             | N                | (Y)        | N                |
| libcxi &  libfabric 1152 & mpich 343 |       NA       | ?               | N                | N          | N                |
| libcxi &  libfabric 1152 & mpich 423 |       NA       | ?               | Y                | N          | N                |
| libcxi &  libfabric 1211 & mpich 423 |       NA       | ?               | N                | Y          | N                |
| libcxi &  libfabric 1220 & mpich 423 |       NA       | ?               | N                | Y          | **Y**            |

> | libcxi &  libfabric 2000 & mpich 423 |       NA       | ?               | ?                | Y          | N                |

Reviewer: What does NA refer to? It should be described in the legend. Or just put an empty space if you apply my suggested change to Y and (Y) in the other comment.

legend:
- Y ==> works at a proper speed
- (Y) ==> Runs but something is slow

> - N ==> Errors out

Reviewer: ```suggestion
- - ==> Errors out
```
As I mentioned I think trying to reduce the letter density of the table makes it easier to digest. I think it also allow you to put actual performance numbers instead of Y and (Y), Perhaps writing 2GBs / 1.5GBs, 22GBs / 13GBs, 0.4GBs / NA etc. and then specify in the legend what the speeds refer to MPI and RCCL communication speeds respectively. If you implement this, you can remove most of the following lists, and just keep with ones that has additional information

- libcxi & libfabric 1152 & mpich 343 --> FAILS; This MPICH doesnt play well with libcxi (open_fabric:No data available)
- libcxi & libfabric 1152 & mpich 423 --> 22GBs; GPU comms work; RCCL much slower (3GBs)
- libcxi & libfabric 1211 & mpich 423 --> FAILS; MPICH expects higher fabric version
- libcxi & libfabric 1220 & mpich 423 --> FAILS; MPICH expects higher fabric version

Opensource:
Notes:
- libfabric 1152 & MPICH 314 -->  WORKS. TCP/IP.
- libfabric 1220 & MPICH 343 --> DOESNT WORK; leads to incompatability with libcxi.
- libfabric 1211 & MPICH 423 -->  WORKS
- libfabric 1220 & MPICH 423 --> WORKS
- libfabric 1152 & MPICH 423 --> DOESNT WORK; libfabric --enable-cxi --> not available
- libcxi & libfabric 1152 & mpich 343 --> FAILS; (open_fabric:No data available)
- libcxi & libfabric 1152 & mpich 423 --> FAILS; (open_fabric:No data available)
- libcxi & libfabric 1211 & mpich 423 --> 22GBs; GPU comms work; RCCL at 13GBs
- libcxi & libfabric 1220 & mpich 423 --> 22GBs; GPU comms work; RCCL at 13GBs
- libcxi & libfabric 2000 & mpich 423 --> 22GBs; GPU comms work; RCCL at 3GBs (**AWS-OFI-RCCL** missing)

> # Definition of basic, lumi bind etc.

Reviewer: As mentioned it would be nice to have this section before the compatibility matrix

Reviewer: Upon having read through it, I think this section seems more like "Implementation details". So I think keeping it here is fine, but write a new "Definitions of container variations" before the compatibility matrix. This definitions chapter should also define the names of the container variations that should be applied throughout.

## Basic
We take the container as is, with MPICH, libfabric etc. 
Nothing gets bind mounted and the communication should run via TCP IP.

The versions used for each library can be reviewed in [Dockerfile.define_versions](common_docker_defs/Dockerfile.define_versions).

This is based on the `base_image_mpich314_libfabric1152.sif` image.

***

## Native

In order to build the native OSU benchmarks you have to run the following:

```
module load LUMI/24.03
module load craype-x86-trento
module load PrgEnv-amd
module load craype-accel-amd-gfx90a
module load rocm
module load EasyBuild-user
export MPICH_GPU_SUPPORT_ENABLED=1
eb aws-ofi-rccl-17d41cb-cpeGNU-24.03.eb -r

> export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/users/$USER/EasyBuild/SW/LUMI-24.03/L/aws-ofi-rccl/17d41cb-cpeGNU-24.03/lib/:/users/$USER/EasyBuild/SW/LUMI-24.03/L/aws-ofi-rccl/17d41cb-cpeGNU-24.03/lib64/

Reviewer: ```suggestion
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$EASYBUILD_PREFIX/SW/LUMI-24.03/L/aws-ofi-rccl/17d41cb-cpeGNU-24.03/lib/:/users/$USER/EasyBuild/SW/LUMI-24.03/L/aws-ofi-rccl/17d41cb-cpeGNU-24.03/lib64/
```

> cd /project/project_XXXXXXXXX/

Reviewer: ```suggestion
mkdir -p /project/project_XXXXXXXXX/$USER
cd /project/project_XXXXXXXXX/$USER
```

curl -LO  https://mvapich.cse.ohio-state.edu/download/mvapich/osu-micro-benchmarks-7.5.tar.gz
tar -xf osu-*.tar.gz 
rm -rf osu-*.tar.gz 
cd osu-*/ 

> CC=cc CXX=CC ./configure --enable-rocm --with-rocm=/opt/rocm-6.0.3 --enable-rcclomb --with-rccl=/opt/rocm-6.0.3 --prefix=/project/project_465001699/osu/build_osu

Reviewer: ```suggestion
CC=cc CXX=CC ./configure --enable-rocm --with-rocm=/opt/rocm-6.0.3 --enable-rcclomb --with-rccl=/opt/rocm-6.0.3 --prefix=/project/project_XXXXXXXXX/$USER/osu/build_osu
```

make install
```

**NOTE:** 
For native bandwidth tests with GPUs, the GTL needs to be linked in, and you need to set 'export MPICH_GPU_SUPPORT_ENABLED=1'.
GTL should be automatically linked when 'module load craype-accel-amd-gfx90a'. 

> Otherwise, you get errors. You can check if the GTL is linked with 'ldd' and the library is 'libmpi_gtl_hsa.so.0 => /opt/cray/pe/lib64/libmpi_gtl_hsa.so.0'.

Reviewer: Where is this library file found after installing natively?

[run_lumi_bind_bandwidth_and_latency_tests.sh](tests/lumi_bind/run_lumi_bind_bandwidth_and_latency_tests.sh)

This is based on the `base_image_mpich314_libfabric1152.sif` image.

The Lumi Full Bind mount is special as it needs to preload the `libhsa-runtime64.so.1` before executing device to device communication. 
Therefore, the container includes the  [run_script.sh](additional_docker_files/run_script.sh) file. This runs the `LD_PRELOAD` command before the actual OSU benchmark. 

### Additional information:

- including 'opt/rocm-6.0.3/lib' in LD_LIBRARY_PATH needed for 'libhsa-runtime64.so.1' 
- LD_LIBRARY_PATH use '/opt/cray/pe/mpich/8.1.29/ofi/amd/5.0/lib-abi-mpich' instead of '/opt/cray/pe/mpich/8.1.29/ofi/gnu/12.3/lib-abi-mpich'
- LD_LIBRARY_PATH needs to include '/opt/rocm-6.0.3/llvm/lib' for 'amd/5.0/lib-abi-mpich'

Additional bind mounts needed over normal singularity bind mounts:
- '/usr/share/libdrm/amdgpu.ids,'
- '/opt/rocm-6.0.3'

I found the correct lib-abi-mpich to include by loading the following:

module load LUMI/24.03
module load craype-x86-trento
module load PrgEnv-amd
module load craype-accel-amd-gfx90a

> module load rocm

Reviewer: ```suggestion
module load LUMI/24.03 craype-x86-trento PrgEnv-amd craype-accel-amd-gfx90a rocm
```
The new-line formatting was off, so we might as well just shorten it.

***
## Libfabric Hybrid
Take a container with libfabric and MPICH and replace libfabric as well as include cxi.
The details are in:
[run_libfabric_hybrid_bandwidth_and_latency_tests.sh](tests/libfabric_hybrid/run_libfabric_hybrid_bandwidth_and_latency_tests.sh)

This is based on the `base_image_libfabric1152_mpich423.sif` image.
The versions of the libraries matter a lot for this way of running containers on LUMI. 

The GTL library is not required for this container as we can use the device communication from MPICH 4.2.2.
However, we do need to enable this with `MPIR_CVAR_CH4_OFI_ENABLE_HMEM=1`

***
## Opensource:
Container with libfabric, MPICH and libcxi. No bind mounts needed.
This is based on the `base_image_libcxi_libfabric1220_mpich423.sif` image.

The GTL library is not required for this container as we can use the device communication from MPICH 4.2.2.
However, we do need to enable this with `MPIR_CVAR_CH4_OFI_ENABLE_HMEM=1`

***
***

# Results

## Basic container

Run with: base_image_mpich314_libfabric1152.sif

- TCP/IP sockets are used for RCCL comms
- MPICH_RELEASE=3.4.3 and LIBFABRIC_RELEASE=1.22.0 lead to very poor performance on normal MPICH comms. About 7x slower. 

> - MPICH version > 4 seems to lead to better results again (but did not check libfabric version)

Reviewer: ```suggestion
- MPICH version > 4 seems to lead to better results again
```
You did check the libfabric version right according to the table right?

| 8192                        | 3.62            | 3.74           | 3.73           | 3.85             | 28.44  |
| 16384                       | 4.25            | 4.39           | 4.35           | 4.43             | 29.33  |
| 32768                       | 7.64            | 7.81           | 7.79           | 7.93             | 34.12  |
| 65536                       | 9.07            | 9.17           | 9.19           | 9.29             | 39.07  |
| 131072                      | 11.97           | 11.99          | 12.00          | 12.07            | 45.44  |
| 262144                      | 17.75           | 17.56          | 17.57          | 17.43            | 57.38  |
| 524288                      | 29.34           | 28.77          | 28.77          | 28.29            | 81.90  |
| 1048576                     | 52.51           | 51.21          | 51.20          | 49.99            | 115.63 |
| 2097152                     | 98.86           | 96.13          | 96.11          | 93.45            | 191.97 |
| 4194304                     | 191.56          | 185.88         | 185.93         | 180.41           | 375.22 |

***
## Libfabric Hybrid

> Run with: base_image_libcxi_libfabric1152_mpich423.sif

Reviewer: This seems to not be one of the default builds, the default libcxi build should be 1220, is this an error?

| 16384                        | 16230.03         | 12800.60       | 17493.16       | 13642.25         | 1139.32  |
| 32768                        | 18746.15         | 16580.44       | 19476.02       | 16534.75         | 2017.60  |
| 65536                        | 20741.79         | 19185.60       | 20928.97       | 20513.81         | 3071.11  |
| 131072                       | 21737.47         | 21627.87       | 21889.16       | 22108.55         | 4980.09  |
| 262144                       | 22184.69         | 22757.22       | 22274.00       | 23007.91         | 6890.97  |
| 524288                       | 22417.82         | 23355.85       | 22448.77       | 23530.24         | 8496.92  |
| 1048576                      | 22536.05         | 23683.30       | 22552.47       | 23742.97         | 11489.69 |
| 2097152                      | 22596.78         | 23837.40       | 22597.25       | 23855.86         | 12618.10 |
| 4194304                      | 22626.56         | 23917.46       | 22620.44       | 23914.85         | 13262.35 |

### Latency

During the running of the latency tests I got the following message multiple times:

> - `srun: Job 9813899 step creation temporarily disabled, retrying (Requested nodes are busy)`

Reviewer: I don't reproduce this warning, however I do get a race condition in e.g. the Open-source rccl latency test, my output starts like:
```
#Using RCCL
#Using RCCL
# Send Buffer on DEVICE (D) and Receive Buffer on DEVICE (D)
# Size          Latency (us)
0                      24.70
1                      49.37
```
I think since the job script asks for 2 nodes and 1 tasks per node you get two mpi processes. Each srun is then probably executed twice simultaneously and this is then done in serial for all srun as the previous one finishes.

| 32768                     | 7.97             | 8.21           | 8.27           | 8.44             | 33.68  |
| 65536                     | 9.68             | 9.57           | 9.61           | 9.79             | 38.63  |
| 131072                    | 12.39            | 12.48          | 12.39          | 12.51            | 45.31  |
| 262144                    | 18.20            | 18.08          | 17.95          | 17.88            | 57.09  |
| 524288                    | 29.81            | 29.17          | 29.27          | 28.73            | 81.68  |
| 1048576                   | 52.89            | 51.86          | 51.55          | 50.26            | 116.05 |
| 2097152                   | 99.13            | 96.48          | 96.74          | 94.21            | 193.36 |
| 4194304                   | 191.76           | 186.25         | 186.09         | 180.67           | 377.31 |

***
***

# Future Work
- GROMACS or similar? https://www.gromacs.org/tutorial_webinar.html
- Speed tests for pytorch
  - real world Pytorch examples would be good
- MPICH with slurm for more options? --> No. MPICH comes with Hydra which supports slurm

> - Can we do better than 25GB/s? 

Reviewer: The Slingshot-11 Interconnect on LUMI is 200 Gbits/s so I don't think you can do better unless you go intra-node.

  - Maybe through some slurm configs to select NICs?
- are the pure RCCL tests necessary?

> - Do we need to do internode testing?

Reviewer: I thought this entire benchmark was internode: In all your script you do --nodes=2 and --gpus-per-node=1. If you had done --nodes=1 and gpus-per-node=2 you would probably see intranode performance which could go up to 200GB/s depending on the NUMA domain.

Author: done
